### ISSUE TITLE
Non-self explanatory Optional property initialization

### ISSUE THREAD

**Issue by raveren** [spatie/laravel-data#583]
**✏️ Describe the bug**

This works:
```php
class SongData extends Data
{
    public string $title;
    public array|Optional $artist;

    public function __construct() {}
}

class DataTest extends TestCase
{
    public function test_case: void
    {
        $s        = SongData::from([]);
        $s->title = 'ads';

        dd($s->toArray());
    }
}
```

This doesn't:

```
Error: Typed property Tests\App\SongData::$artist must not be accessed before initialization
/var/www/vendor/spatie/laravel-data/src/Transformers/DataTransformer.php:79
/var/www/vendor/spatie/laravel-data/src/Transformers/DataTransformer.php:48
/var/www/vendor/spatie/laravel-data/src/Concerns/BaseData.php:106
/var/www/vendor/spatie/laravel-data/src/Concerns/TransformableData.php:16
```

```php
class SongData extends Data
{
    public string $title;
    public array|Optional $artist;

    public function __construct() {}
}

class DataTest extends TestCase
{
    public function test_case: void
    {
        $s        = SongData::from(); // <-- removed empty array
        $s->title = 'ads';

        dd($s->toArray());
    }
}
```

And most importantly this, natural use, which would allow usage of LaravelData without any documentation doesn't work either:

```php
class SongData extends Data
{
    public string $title;
    public array|Optional $artist;

    public function __construct() {}
}

class DataTest extends TestCase
{
    public function test_case: void
    {
        $s        = new SongData(); // <-- 
        $s->title = 'ads';

        dd($s->toArray());
    }
}
```

**✅ Expected behavior**

Logically, the class initialization should not alter this behaviour, the initialization should yield equivalent results in the end.

Especially the `DataClass::from()` vs `DataClass::from([])` is a very niche solution for this, feels like a work-around.

**🖥️ Versions**

Laravel: latest
Laravel Data: latest
PHP: latest

**Comment by rubenvanassche** [comment]
Yeah weird behavior, fixed it in the version released today.

**Comment by rubenvanassche** [comment]
Reverted due to the implementation being more difficult than expected: #590

**Comment by ronssij** [comment]
@raveren from what I have understand if the field is optional, sometimes it could be null and it is a realtionship. you can do the `Data|Lazy|Optional`

```
class SongData extends Data
{
    public string $title;
    public ArtistData|Lazy|Optional $artist;

    public function __construct() {}
}
```

**Comment by raveren** [comment]
@ronssij I don't fully understand what you mean, I think you suggest using Lazy as another workaround..?

However it is counter-intuitive to come up with and to maintain, so not really a solution. And it WILL have other side-effects down the road. 

It's on the same level as calling `MyClass::from([])` just an another step further from the truth.

For now I am sticking with 3.9.1, this challenge is huge and the implications when you guys do it right will be huge as well, so I'm patient, working around, waiting, wishing you luck and determination @rubenvanassche & co!

**Comment by ronssij** [comment]
![image](https://github.com/spatie/laravel-data/assets/24932380/dec54150-de4e-458e-89b7-b6fab105dd52)

This is my sample working data.

![image](https://github.com/spatie/laravel-data/assets/24932380/6ab1405d-3eed-4d7d-af87-5207901c303a)

**Comment by rubenvanassche** [comment]
Should be fixed with #602
